File: kelly/BFS & DFS/Keys and Rooms/solution02.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False] * len(rooms)
        stack = [0]
        visited[0] = True

        while stack:
            cur = stack.pop()
            for next in rooms[cur]:
                if visited[next]:
                    continue

                stack.append(next)
                visited[next] = True

        return not False in visited

File: kelly/BFS & DFS/Keys and Rooms/Solution02.java
import java.util.*;

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0);
        visited[0] = true;

        while (!stack.isEmpty()) {
            int cur = stack.pop();

            for (int next : rooms.get(cur)) {
                if (visited[next]) {
                    continue;
                }

                stack.push(next);
                visited[next] = true;
            }
        }

        for (boolean v : visited) {
            if (!v) {
                return false;
            }
        }

        return true;
    }
}